### PR TITLE
Feature/adapter deployment

### PR DESCRIPTION
**Pull request by anwar6953**
Summary of PR:
- Return 409 if an adapter-create request occurs, and an adapter already exists with the same namespace + adapterName.
- Add status endpoint (storing status in AdapterMeta in mds - instead of storing appSpec directly in mds).
- Relevant test cases / doc updates

Build: https://builds.cask.co/browse/CDAP-RBT61-13/log

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 1]
```diff
@@ -350,6 +352,23 @@ public void startStopAdapter(HttpRequest request, HttpResponder responder,
...
+    } catch (AdapterConflictException e) {
+      responder.sendString(HttpResponseStatus.CONFLICT, e.getMessage());
+    }
+  }
+
+  /**
+   * Retrieves the status of an adapter
+   */
+  @GET
+  @Path("/adapters/{adapterId}/status")
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: ```{adapter-id}``` to make it consistent.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 2]
```diff
@@ -332,29 +332,29 @@ void addApplication(Id.Application id,
...
 
   /**
    * Fetch all the adapters in a given namespace.
    *
    * @param id Namespace id.
-   * @return {@link Collection} of Adapter Specification.
+   * @return {@link Collection} of Adapter Meta.
    */
-  Collection<AdapterSpecification> getAllAdapters(Id.Namespace id);
+  Collection<AdapterMeta> getAllAdapters(Id.Namespace id);
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: We should not change this API. ```getAllAdapters``` should return the specification and not the runtime state. All our other APIs donot mix specification with runtime state. I would recommend adding a ```status``` API to ```AdapterService``` instead to retrieve this information

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterAlreadyExistsException.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.adapter;
+
+/**
+ * Thrown when Adapter creation is requested, but an adapter already exists.
+ */
+public class AdapterAlreadyExistsException extends Exception {
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterAlreadyExistsException.java]: We create an Exception here and also another one in client. We try to consolidate it. We should move everything to common module. We can then throw generic ```AlreadyExistsException```

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterConflictException.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.adapter;
+
+/**
+ * Thrown when Adapter conflict occurs. For instance, if an adapter 'stop' is requested, but the adapter is already
+ * stopped.
+ */
+public class AdapterConflictException extends Exception {
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterConflictException.java]: Conflict is a state, the actual issue is that you cannot perform the operation expected. How about ```InvalidOperationException``` and then we can return a ```Conflict - 409``` from the HTTP layer that calls it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java [hunk 5]
```diff
@@ -134,11 +135,23 @@ public AdapterTypeInfo getAdapterTypeInfo(String adapterType) {
...
+
+  /**
+   * Retrieves the {@link AdapterSpecification} specified by the name in a given namespace.
+   *
+   * @param namespace namespace to lookup the adapter
+   * @param adapterName name of the adapter
+   * @return requested {@link AdapterSpecification} or null if no such AdapterInfo exists
+   * @throws AdapterNotFoundException if the requested adapter is not found
+   */
+  public AdapterMeta getAdapterMeta(String namespace, String adapterName) throws AdapterNotFoundException {
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java]: We should not expose this API. Remove it.

**Comment by sreevatsanraman** [comment]
Please fix the initial set of comments. Will take another pass at it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 6]
```diff
@@ -349,10 +351,29 @@ void addApplication(Id.Application id,
   AdapterSpecification getAdapter(Id.Namespace id, String name);
 
   /**
+   * Fetch the status for an adapter identified by the name in a give namespace.
+   *
+   * @param id  Namespace id.
+   * @param name Adapter name
+   * @return status of specified adapter.
+   */
+  String getAdapterStatus(Id.Namespace id, String name);
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: ```@Nullable```

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 7]
```diff
@@ -349,10 +351,29 @@ void addApplication(Id.Application id,
...
+
+  /**
+   * Set the status for an adapter identified by the name in a give namespace.
+   *
+   * @param id  Namespace id.
+   * @param name Adapter name
+   * @param status Status to set
+   * @return previous status of adapter, or null if specified adapter is not found.
+   */
+  String setAdapterStatus(Id.Namespace id, String name, String status);
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: Nullable

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java [hunk 8]
```diff
@@ -227,10 +268,18 @@ public void stopAdapter(String namespace, String adapterName) throws AdapterNotF
...
+    setAdapterStatus(namespace, adapterName, AdapterStatus.STOPPED);
   }
 
   // Resumes all schedules for this adapter
-  public void startAdapter(String namespace, String adapterName) throws AdapterNotFoundException {
+  public void startAdapter(String namespace, String adapterName)
+    throws AdapterNotFoundException, InvalidAdapterOperationException {
+    String adapterStatus = getAdapterStatus(namespace, adapterName);
+    if (AdapterStatus.STARTED.equals(adapterStatus)) {
+      throw new InvalidAdapterOperationException("Adapter is already stopped.");
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java]: started

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterStatus.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.adapter;
+
+/**
+ * Possible status of an Adapter
+ */
+public class AdapterStatus {
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterStatus.java]: use an enum

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterLifecycleTests.java [hunk 10]
```diff
@@ -104,6 +109,30 @@ public void testAdapterLifeCycle() throws Exception {
     JsonObject deployedApp = deployedApps.get(0);
     Assert.assertEquals(adapterType, deployedApp.get("id").getAsString());
 
+    response = getAdapterStatus(namespaceId, adapterName);
+    Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+    String status = readResponse(response, String.class);
+    Assert.assertEquals("STARTED", status);
+
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterLifecycleTests.java]: Add tests to check conflicts - 409.

**Comment by sreevatsanraman** [comment]
Minor comments, fix and merge. Open a JIRA for consolidating the Exception classes between client and rest of the code.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterLifecycleTests.java (hunk 10 above)
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterLifecycleTests.java]: already exist. line 125 of AdapterLifecycleTests.java

**Comment by anwar6953** [comment]
https://issues.cask.co/browse/CDAP-1257

**Comment by sreevatsanraman** [comment]
:+1:

**Pull request merged**